src/darwin/Framework/CHIP/MTRDeviceControllerFactory.mm

```diff
@@ -452,47 +395,44 @@ - (BOOL)_startControllerFactory:(MTRDeviceControllerFactoryParams *)startupParam
         _controllerFactory->RetainSystemState();
         _controllerFactory->ReleaseSystemState();
 
-        self->_advertiseOperational = startupParams.shouldStartServer;
-        self->_running = YES;
-    });
+        _advertiseOperational = startupParams.shouldStartServer;
+        _running = YES;
+        err = CHIP_NO_ERROR;
 
-    if (![self isRunning]) {
-        dispatch_sync(_chipWorkQueue, ^{
+    exit:
+        if (err != CHIP_NO_ERROR) {
+            // Note: Since we have failed no later than _controllerFactory->Init(),
+            // there is no need to call _controllerFactory->Shutdown() here.
             [self cleanupStartupObjects];
-        });
+        }
+    });
+    if (err != CHIP_NO_ERROR) {
+        MTR_LOG_ERROR("Failed to start Matter controller factory: %" CHIP_ERROR_FORMAT, err.Format());
         if (error != nil) {
-            *error = [MTRError errorForCHIPErrorCode:errorCode];
+            *error = [MTRError errorForCHIPErrorCode:err];
         }
         return NO;
     }
-
     return YES;
 }
 
 - (void)stopControllerFactory
 {
     [self _assertCurrentQueueIsNotMatterQueue];
 
-    if (![self isRunning]) {
```

Comment:
Is this safe to remove?

Reply:
I'm removing a couple of the double checks if favor of checking just within the dispatch_sync. We want to work towards having the factory API be fully thread-safe (currently it assumes some level of external synchronization for some operations.)